rules/TypeDeclaration/TypeInferer/ReturnTypeInferer/YieldNodesReturnTypeInfererTypeInferer.php

```diff
@@ -51,19 +60,44 @@ public function inferFunctionLike(FunctionLike $functionLike): Type
             $types[] = $resolvedType;
         }
 
+        $className = $this->resolveClassNameReturnType($functionLike);
+
         if ($types === []) {
-            return new FullyQualifiedObjectType('Iterator');
+            return new FullyQualifiedObjectType($className);
         }
 
         $types = $this->typeFactory->createMixedPassedOrUnionType($types);
-        return new FullyQualifiedGenericObjectType('Iterator', [$types]);
+        return new FullyQualifiedGenericObjectType($className, [$types]);
     }
 
     public function getPriority(): int
     {
         return 1200;
     }
 
+    private function resolveClassNameReturnType(FunctionLike $functionLike): string
+    {
+        if (! $functionLike->getReturnType() instanceof Node) {
+            $phpDocInfo = $this->phpDocInfoFactory->createFromNodeOrEmpty($functionLike);
+            $returnTagValueNode = $phpDocInfo->getReturnTagValue();
+
+            if (! $returnTagValueNode instanceof ReturnTagValueNode) {
+                return self::CLASS_NAME;
+            }
+
+            if (! $returnTagValueNode->type instanceof GenericTypeNode) {
+                return self::CLASS_NAME;
+            }
+
+            $typeName = ltrim($returnTagValueNode->type->type->name, '\\');
+            if ($typeName === 'Generator') {
+                return 'Generator';
+            }
+        }
+
+        return self::CLASS_NAME;
```

Comment:
This looks like hardcoded resolution from the docblock. What if there is wrong type in `@return`?

Reply:
wrong type in `@return` will fallback to `Iterator` as constant `self::CLASS_NAME` defined.

Reply:
I added new fixture for wrong type generator via `@return`, it goes to `Iterator` return type https://github.com/rectorphp/rector-src/pull/1794/commits/d18bd5a2089e1c5f4abf76a811016940fd11800e